src/jsCodegen/compileAstToJavaScript.ts

```diff
@@ -0,0 +1,127 @@
+import { ValueType } from '../expressions/dataTypes/Value';
+import astHelper, { IAST } from '../parsing/astHelper';
+import { ReturnType } from '../parsing/convertXDMReturnValue';
+import { emitBaseExpr } from './emitBaseExpression';
+import {
+	acceptAst,
+	acceptFullyCompiledAst,
+	JavaScriptCompiledXPathResult,
+	PartialCompilationResult,
+	PartiallyCompiledAstAccepted,
+	rejectAst,
+} from './JavaScriptCompiledXPath';
+import { StaticContext } from './StaticContext';
+
+// Return all matching nodes.
+function emitEvaluationToNodes(identifier: string): PartiallyCompiledAstAccepted {
+	return acceptAst(`
+	const nodes = [];
+	for (const node of ${identifier}(contextItem)) {
+		nodes.push(node.value.node);
+	}
+	return nodes;
+	`);
+}
+
+// Get effective boolean value.
+function emitEvaluationToBoolean(identifier: string): PartiallyCompiledAstAccepted {
+	return acceptAst(`
+	return determinePredicateTruthValue(${identifier}(contextItem));
+	`);
+}
+
+function emitEvaluationToFirstNode(identifier: string): PartiallyCompiledAstAccepted {
+	return acceptAst(`
+	return ${identifier}(contextItem).next().value.value.node
+	`);
+}
+
+function emitReturnTypeConversion(
+	identifier: string,
+	returnType: ReturnType
+): PartialCompilationResult {
+	switch (returnType) {
+		case ReturnType.FIRST_NODE:
+			return emitEvaluationToFirstNode(identifier);
+		case ReturnType.NODES:
+			return emitEvaluationToNodes(identifier);
+		case ReturnType.BOOLEAN:
+			return emitEvaluationToBoolean(identifier);
+		default:
+			return rejectAst(`Unsupported: the return type '${returnType}'.`);
+	}
+}
+
+function wrapCompiledCode(code: string): string {
+	return `
+	return (contextItem, domFacade, runtimeLibrary) => {
+		const {
+			DONE_TOKEN,
+			ValueType,
+			XPDY0002,
+			adaptSingleJavaScriptValue,
+			determinePredicateTruthValue,
+			isSubtypeOf,
+			ready,
+		} = runtimeLibrary;
```

Comment:
I am actually surprised this works in the closure-compiled version of fontoxpath. I expected all these names to be renamed to `a` or `$`. Can you verify this all works in the bundle as well?

Reply:
You can verify this by including the jsCodeGen tests in the `npm run integrationtests` script.